driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,401 @@
+
+        private ErrorAwarePolicyTracker errorTracker;
+        private long updatePeriod = 1000;
+        private int errorsPerSecThreshold = 10;
+        private int retryPeriod = 5000;
+
+        public Builder(LoadBalancingPolicy childPolicy) {
+            this.childPolicy = childPolicy;
+        }
+
+        public Builder withUpdatePeriodMillis(long updatePeriodMillis) {
+            this.updatePeriod = updatePeriodMillis;
+            return this;
+        }
+
+        // Hosts that have a number of errors/sec higher than the threshold are ignored.
+        public Builder withInclusionThreshold(int errorsPerSecThreshold) {
+            this.errorsPerSecThreshold = errorsPerSecThreshold;
+            return this;
+        }
+
+        public Builder withErrorTracker(ErrorAwarePolicyTracker errorTracker) {
+            this.errorTracker = errorTracker;
+            return this;
+        }
+
+        public Builder withRetryPeriod(int retryPeriod) {
+            this.retryPeriod = retryPeriod;
+            return this;
+        }
+
+        public ErrorAwarePolicy build() {
+            return new ErrorAwarePolicy(childPolicy, errorTracker, updatePeriod, errorsPerSecThreshold, retryPeriod);
+        }
+    }
+
+    public static class PerHostErrorTracker implements ErrorAwarePolicyTracker {
+
+        private Map<Host, Integer> errors = new HashMap<Host, Integer>();
+
+        private Map<Host, ErrorRecord> errorsPerSec = new HashMap<Host, ErrorRecord>();
+
+        private long retryPeriod;
+
+        private int errorsPerSecThreshold;
+
+        private double errorsThresholdRate;
+
+        private long updatePeriod;
+
+        @Override
+        public void notifyError(Host host, Statement statement, Exception exception) {
+            // should just keep counts of errors for each hosts
+            if (!shouldAccountForError(host, statement, exception)) {
+                return;
+            }
+            Integer entry;
+            Integer count = errors.get(host);
+            if (count == null) {
+                entry = 1;
+            } else {
+                entry = count + 1;
+            }
+            errors.put(host, entry);
+        }
+
+        @Override
+        public void onRegister(Cluster cluster) {
+
+        }
+
+        @Override
+        public void onUnregister(Cluster cluster) {
+
+        }
+
+        @Override
+        public void update() {
+            Map<Host, Integer> errorsSinceLastUpdate = new HashMap<Host, Integer>(errors.size());
+
+            // Update the error rates
+            for (Map.Entry<Host, Integer> errorsByHost : errors.entrySet()) {
+                int errorsNb = errorsByHost.getValue();
+                // TODO: we may miss errors between get and put...
+                errors.put(errorsByHost.getKey(), 0);
+                errorsSinceLastUpdate.put(errorsByHost.getKey(), errorsNb);
+
+                double errorsPerSecHostRate = ((double) (errorsNb * updatePeriod)) / 1000;
```

Comment:
Ok I've put a little more thoughts into it, when I wrote the code I imagined the _errorsRate_ to be representation of the number of errors at the scale of the period. With a simple example of updatePeriod = 100 (ms) and threshold = 2 (errors/sec) :
(so putting aside my false maths simplification tries) Here, since the updatePeriod is **smaller than a second**, a calculated error rate, for me, should look like a "less important" representation of the errors happening, with regards to the threshold specified in errors per **seconds**. I.e. with `errorsNb / (1000 / updatePeriod)`, 1 error happening in a 100ms frame would make the errorRate = 0.1. Whereas if you